src/registry.rs

```diff
@@ -0,0 +1,164 @@
+use std::sync::atomic::{AtomicUsize, ATOMIC_USIZE_INIT};
+use std::sync::atomic::Ordering::{AcqRel, Acquire, Relaxed, Release};
+
+use {Atomic, Owned, Ptr, Scope, unprotected};
+use participant::Participant;
+
+// FIXME(stjepang): Participants are stored in a linked list because linked lists are fairly easy
+// to implement in a lock-free manner. However, traversal is rather slow due to cache misses and
+// data dependencies. We should experiment with other data structures as well.
+
+/// An entry in the linked list of participating threads.
+struct Entry {
+    /// State of the participating thread.
+    participant: Participant,
+
+    /// The next entry in the linked list.
+    /// If the tag is 1, this entry is marked as deleted.
+    next: Atomic<Entry>,
+}
+
+/// Registers a thread by inserting a new entry into the list of participating threads.
+///
+/// Returns a pointer to the newly allocated and inserted entry.
+fn register() -> *const Entry {
+    let list = participants();
+
+    let mut new = Owned::new(Entry {
+        participant: Participant::new(),
+        next: Atomic::null(),
+    });
+
+    unsafe {
+        // Since we don't dereference any pointers in this block, it's okay to use `unprotected`.
+        unprotected(|scope| {
+            let mut head = list.load(Acquire, scope);
+
+            loop {
+                new.next.store(head, Relaxed);
+
+                // Try installing the new entry as the new head.
+                match list.compare_and_set_weak_owned(head, new, AcqRel, scope) {
+                    Ok(n) => return n.as_raw(),
+                    Err((h, n)) => {
+                        head = h;
+                        new = n;
+                    }
+                }
+            }
+        })
+    }
+}
+
+thread_local! {
+    /// The thread registration harness.
+    ///
+    /// The harness is lazily initialized on its first use, thus registrating the current thread.
```

Comment:
Yes, according to the docs:
https://doc.rust-lang.org/std/thread/struct.LocalKey.html#initialization-and-destruction